src/flags.zig

```diff
@@ -302,84 +302,158 @@ fn parse_value(comptime T: type, flag: []const u8, value: [:0]const u8) T {
             ),
-            error.InvalidCharacter => fatal(
-                "{s}: expected a size, but found '{s}' (invalid digit or suffix)",
+            error.InvalidSize => fatal(
+                "{s}: expected a size, but found '{s}'",
+                .{ flag, value },
+            ),
+            error.InvalidUnit => fatal(
+                "{s}: invalid unit in size '{s}', (needed KiB, MiB, GiB or TiB)",
+                .{ flag, value },
+            ),
+            error.BytesOverflow => fatal(
+                "{s}: size in bytes exceeds 64-bit unsigned integer: '{s}'",
                 .{ flag, value },
             ),
         }
     };
-
-    const bytes = std.math.mul(u64, amount, unit.scale) catch fatal(
-        "{s}: size in bytes exceeds 64-bit unsigned integer: '{s}'",
-        .{ flag, value },
-    );
-    return ByteSize{ .bytes = bytes };
 }
 
+pub const ByteUnit = enum(u64) {
+    bytes = 1,
+    kib = 1024,
+    mib = 1024 * 1024,
+    gib = 1024 * 1024 * 1024,
+    tib = 1024 * 1024 * 1024 * 1024,
+};
+
+const ByteSizeParseError = error{
+    ParseOverflow,
+    InvalidSize,
+    InvalidUnit,
+    BytesOverflow,
+};
+
+pub const ByteSize = struct {
+    value: u64,
+    unit: ByteUnit = .bytes,
+
+    fn parse(value: []const u8) ByteSizeParseError!ByteSize {
+        assert(value.len != 0);
+
+        const units = .{
+            .{ "TiB", ByteUnit.tib },
+            .{ "GiB", ByteUnit.gib },
+            .{ "MiB", ByteUnit.mib },
+            .{ "KiB", ByteUnit.kib },
+        };
+
+        const split: struct {
+            value_input: []const u8,
+            unit_input: []const u8,
+        } = split: for (0..value.len) |i| {
+            if (!std.ascii.isDigit(value[i])) {
+                break :split .{
+                    .value_input = value[0..i],
+                    .unit_input = value[i..],
+                };
+            }
+        } else {
+            break :split .{
+                .value_input = value,
+                .unit_input = "",
+            };
+        };
+
+        const amount = std.fmt.parseUnsigned(u64, split.value_input, 10) catch |err| {
+            switch (err) {
+                error.Overflow => {
+                    return ByteSizeParseError.ParseOverflow;
+                },
+                error.InvalidCharacter => {
+                    // The only case this can happen is for the empty string
+                    return ByteSizeParseError.InvalidSize;
+                },
+            }
+        };
+
+        const unit = if (split.unit_input.len > 0)
+            unit: inline for (units) |unit_| {
+                const suffix_ = unit_[0];
+                const unit_kind = unit_[1];
+                if (std.ascii.eqlIgnoreCase(split.unit_input, suffix_)) {
+                    break :unit unit_kind;
+                }
+            } else {
+                return ByteSizeParseError.InvalidUnit;
+            }
+        else
+            ByteUnit.bytes;
```

Comment:
Even better, we don't need the `units` tuples mapping text to enums anymore! 

```suggestion
        const unit = if (split.unit_input.len > 0)
            unit: inline for (std.enums.values(ByteUnit)) |tag| {
                if (std.ascii.eqlIgnoreCase(split.unit_input, @tagName(tag))) {
                    break :unit tag;
                }
            } else {
                return ByteSizeParseError.InvalidUnit;
            };
        else
            ByteUnit.bytes;
```

Reply:
Done.